File: be/src/http/action/proc_profile_action.cpp
// Copyright 2021-present StarRocks, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "http/action/proc_profile_action.h"

#include <filesystem>
#include <string>
#include <vector>

#include "common/config.h"
#include "common/logging.h"
#include "gutil/strings/substitute.h"
#include "http/action/profile_utils.h"
#include "http/http_channel.h"
#include "http/http_headers.h"
#include "http/http_method.h"
#include "http/http_request.h"
#include "http/http_status.h"
#include "rapidjson/document.h"
#include "rapidjson/prettywriter.h"
#include "rapidjson/stringbuffer.h"

namespace starrocks {

const static std::string HEADER_JSON = "application/json";
const static std::string ACTION_KEY = "action";
const static std::string ACTION_LIST = "list";

void ProcProfileAction::handle(HttpRequest* req) {
    VLOG_ROW << req->debug_string();
    const auto& action = req->param(ACTION_KEY);
    if (req->method() == HttpMethod::GET) {
        if (action == ACTION_LIST) {
            _handle_list(req);
        } else {
            _handle_error(req, strings::Substitute("Not support action: '$0'", action));
        }
    } else {
        _handle_error(req,
                      strings::Substitute("Not support $0 method: '$1'", to_method_desc(req->method()), req->uri()));
    }
}

void ProcProfileAction::_handle_list(HttpRequest* req) {
    rapidjson::Document root;
    root.SetObject();
    rapidjson::Document::AllocatorType& allocator = root.GetAllocator();

    rapidjson::Value profiles(rapidjson::kArrayType);

    std::string profile_log_dir = std::string(config::sys_log_dir) + "/proc_profile";
    std::filesystem::path dir_path(profile_log_dir);

    if (!std::filesystem::exists(dir_path) || !std::filesystem::is_directory(dir_path)) {
        root.AddMember("profiles", profiles, allocator);
        rapidjson::StringBuffer strbuf;
        rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(strbuf);
        root.Accept(writer);
        req->add_output_header(HttpHeaders::CONTENT_TYPE, HEADER_JSON.c_str());
        HttpChannel::send_reply(req, HttpStatus::OK, strbuf.GetString());
        return;
    }

    std::vector<std::pair<std::string, std::string>> profile_files; // filename, timestamp

    try {
        for (const auto& entry : std::filesystem::directory_iterator(dir_path)) {
            if (entry.is_regular_file()) {
                std::string filename = entry.path().filename().string();
                // Only include files in the pprof format
                if (ProfileUtils::get_profile_format(filename) == "Pprof") {
                    // Extract timestamp from filename using utility function
                    std::string timestamp = ProfileUtils::extract_timestamp_from_filename(filename);
                    if (!timestamp.empty()) {
                        profile_files.emplace_back(filename, timestamp);
                    }
                }
            }
        }
    } catch (const std::filesystem::filesystem_error& e) {
        _handle_error(req, strings::Substitute("Error reading profile directory: $0", e.what()));
        return;
    }

    // Sort by timestamp descending (newest first)
    std::sort(profile_files.begin(), profile_files.end(),
              [](const std::pair<std::string, std::string>& a, const std::pair<std::string, std::string>& b) {
                  return a.second > b.second;
              });

    for (const auto& file_info : profile_files) {
        const std::string& filename = file_info.first;
        const std::string& timestamp = file_info.second;

        // Determine profile type and format using utility functions
        std::string profile_type = ProfileUtils::get_profile_type(filename);
        std::string format = ProfileUtils::get_profile_format(filename);

        // Get file size
        std::string file_path = profile_log_dir + "/" + filename;
        uint64_t file_size = 0;
        try {
            file_size = std::filesystem::file_size(file_path);
        } catch (const std::filesystem::filesystem_error&) {
            // Keep 0 if we can't get file size
        }

        rapidjson::Value profile(rapidjson::kObjectType);
        profile.AddMember("filename", rapidjson::Value(filename.c_str(), allocator), allocator);
        profile.AddMember("type", rapidjson::Value(profile_type.c_str(), allocator), allocator);
        profile.AddMember("format", rapidjson::Value(format.c_str(), allocator), allocator);
        profile.AddMember("timestamp", rapidjson::Value(timestamp.c_str(), allocator), allocator);
        profile.AddMember("size", file_size, allocator);

        profiles.PushBack(profile, allocator);
    }

    root.AddMember("profiles", profiles, allocator);

    rapidjson::StringBuffer strbuf;
    rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(strbuf);
    root.Accept(writer);
    req->add_output_header(HttpHeaders::CONTENT_TYPE, HEADER_JSON.c_str());
    HttpChannel::send_reply(req, HttpStatus::OK, strbuf.GetString());
}

void ProcProfileAction::_handle_error(HttpRequest* req, const std::string& error_msg) {
    rapidjson::Document root;
    root.SetObject();
    rapidjson::Document::AllocatorType& allocator = root.GetAllocator();
    root.AddMember("error", rapidjson::Value(error_msg.c_str(), allocator), allocator);

    rapidjson::StringBuffer strbuf;
    rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(strbuf);
    root.Accept(writer);
    req->add_output_header(HttpHeaders::CONTENT_TYPE, HEADER_JSON.c_str());
    HttpChannel::send_reply(req, HttpStatus::INTERNAL_SERVER_ERROR, strbuf.GetString());
}

} // end namespace starrocks

File: fe/fe-core/src/main/java/com/starrocks/http/action/ProcProfileAction.java
// Copyright 2021-present StarRocks, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.starrocks.http.action;

import com.google.common.collect.ImmutableMap;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.starrocks.common.Config;
import com.starrocks.common.util.DebugUtil;
import com.starrocks.http.ActionController;
import com.starrocks.http.BaseRequest;
import com.starrocks.http.BaseResponse;
import com.starrocks.http.HttpUtils;
import com.starrocks.http.IllegalArgException;
import com.starrocks.server.GlobalStateMgr;
import com.starrocks.system.Backend;
import io.netty.handler.codec.http.HttpMethod;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.File;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class ProcProfileAction extends WebBaseAction {
    private static final Logger LOG = LogManager.getLogger(ProcProfileAction.class);

    private static final String CPU_FILE_NAME_PREFIX = "cpu-profile-";
    private static final String MEM_FILE_NAME_PREFIX = "mem-profile-";
    private static final SimpleDateFormat PROFILE_TIME_FORMAT = new SimpleDateFormat("yyyyMMdd-HHmmss");

    public ProcProfileAction(ActionController controller) {
        super(controller);
    }

    public static void registerAction(ActionController controller) throws IllegalArgException {
        controller.registerHandler(HttpMethod.GET, "/proc_profile", new ProcProfileAction(controller));
    }

    @Override
    public void executeGet(BaseRequest request, BaseResponse response) {
        getPageHeader(request, response.getContent());

        // Add proc profile tabs CSS link in the head section
        // Insert before the closing </head> tag by replacing the body tag opening
        String content = response.getContent().toString();
        int headEndIndex = content.lastIndexOf("</head>");
        if (headEndIndex > 0) {
            response.getContent().setLength(0);
            response.getContent().append(content.substring(0, headEndIndex));
            response.getContent()
                    .append("  <link href=\"/static/css?res=proc-profile-tabs.css\" ")
                    .append("rel=\"stylesheet\" media=\"screen\"/>\n");
            response.getContent().append(content.substring(headEndIndex));
        }

        String nodeParam = request.getSingleParameter("node");
        if (nodeParam == null || nodeParam.isEmpty()) {
            nodeParam = "FE";
        }

        addTabNavigation(response.getContent(), nodeParam);
        addProfileListInfo(response.getContent(), nodeParam);

        getPageFooter(response.getContent());

        writeResponse(request, response);
    }

    private void addTabNavigation(StringBuilder buffer, String selectedNode) {
        buffer.append("<ul class=\"nav nav-tabs\" style=\"margin-bottom: 20px;\">");

        // FE tab
        String feClass = "FE".equals(selectedNode) ? "active" : "";
        buffer.append("<li class=\"").append(feClass).append("\">");
        buffer.append("<a href=\"/proc_profile?node=FE\">FE</a>");
        buffer.append("</li>");

        // BE tabs
        ImmutableMap<Long, Backend> backendMap =
                GlobalStateMgr.getCurrentState().getNodeMgr().getClusterInfo().getIdToBackend();
        for (Backend backend : backendMap.values()) {
            if (!backend.isAlive()) {
                continue;
            }
            String beNodeId = "BE:" + backend.getHost() + ":" + backend.getHttpPort();
            String beClass = beNodeId.equals(selectedNode) ? "active" : "";
            buffer.append("<li class=\"").append(beClass).append("\">");
            buffer.append("<a href=\"/proc_profile?node=").append(beNodeId).append("\">");
            buffer.append("BE: ").append(backend.getHost()).append(":").append(backend.getHttpPort());
            buffer.append("</a>");
            buffer.append("</li>");
        }

        buffer.append("</ul>");
    }

    private void addProfileListInfo(StringBuilder buffer, String nodeParam) {
        buffer.append("<h2>Process Profiles</h2>");
        buffer.append("<p>This table lists all available CPU and memory profile files</p>");

        List<ProfileFileInfo> profileFiles;
        if ("FE".equals(nodeParam)) {
            profileFiles = getProfileFiles();
        } else if (nodeParam.startsWith("BE:")) {
            profileFiles = getProfileFilesFromBE(nodeParam);
        } else {
            buffer.append("<p>Invalid node parameter: ").append(nodeParam).append("</p>");
            return;
        }

        if (profileFiles.isEmpty()) {
            if ("FE".equals(nodeParam)) {
                buffer.append("<p>No profile files found in directory: ")
                        .append(Config.sys_log_dir)
                        .append("/proc_profile</p>");
            } else {
                buffer.append("<p>No profile files found or node unavailable</p>");
            }
            return;
        }

        appendProfileTableHeader(buffer);
        appendProfileTableBody(buffer, profileFiles, nodeParam);
        appendTableFooter(buffer);
    }

    private List<ProfileFileInfo> getProfileFiles() {
        List<ProfileFileInfo> profileFiles = new ArrayList<>();
        String profileLogDir = Config.sys_log_dir + "/proc_profile";
        File dir = new File(profileLogDir);

        if (!dir.exists() || !dir.isDirectory()) {
            LOG.warn("Profile directory does not exist: {}", profileLogDir);
            return profileFiles;
        }

        File[] files = dir.listFiles();
        assert files != null;
        for (File file : files) {
            if (file.isFile() && file.getName().endsWith(".tar.gz")) {
                String fileName = file.getName();
                String profileType = getProfileType(fileName);
                if (profileType != null) {
                    String timePart = getTimePart(fileName);
                    if (timePart != null) {
                        try {
                            Date timestamp = PROFILE_TIME_FORMAT.parse(timePart);
                            profileFiles.add(new ProfileFileInfo(
                                    profileType,
                                    timestamp,
                                    file.length(),
                                    fileName
                            ));
                        } catch (Exception e) {
                            LOG.warn("Failed to parse timestamp from filename: {}", fileName, e);
                        }
                    }
                }
            }
        }

        // Sort by timestamp descending (newest first)
        profileFiles.sort((f1, f2) -> f2.timestamp.compareTo(f1.timestamp));

        return profileFiles;
    }

    private String getProfileType(String fileName) {
        if (fileName.startsWith(CPU_FILE_NAME_PREFIX)) {
            return "CPU";
        } else if (fileName.startsWith(MEM_FILE_NAME_PREFIX)) {
            return "Memory";
        }
        return null;
    }

    private String getTimePart(String fileName) {
        String prefix = fileName.startsWith(CPU_FILE_NAME_PREFIX) ? CPU_FILE_NAME_PREFIX : MEM_FILE_NAME_PREFIX;
        int startIndex = prefix.length();
        int endIndex = fileName.indexOf(".html.tar.gz");
        if (endIndex > startIndex) {
            return fileName.substring(startIndex, endIndex);
        }
        return null;
    }

    private void appendProfileTableHeader(StringBuilder buffer) {
        buffer.append(
                "<table id=\"table_id\" class=\"table table-hover table-bordered table-striped table-condensed\">");
        buffer.append("<thead><tr>");
        buffer.append("<th>Type</th>");
        buffer.append("<th>Timestamp</th>");
        buffer.append("<th>File Size</th>");
        buffer.append("<th>View</th>");
        buffer.append("</tr></thead>");
    }

    private List<ProfileFileInfo> getProfileFilesFromBE(String beNodeId) {
        List<ProfileFileInfo> profileFiles = new ArrayList<>();

        // Parse BE node: "BE:host:port"
        String[] parts = beNodeId.substring(3).split(":");
        if (parts.length != 2) {
            LOG.warn("Invalid BE node format: {}", beNodeId);
            return profileFiles;
        }

        String host = parts[0];
        int port;
        try {
            port = Integer.parseInt(parts[1]);
        } catch (NumberFormatException e) {
            LOG.warn("Invalid BE port: {}", parts[1]);
            return profileFiles;
        }

        try {
            String url = "http://" + host + ":" + port + "/api/proc_profile/list";
            String jsonResponse = HttpUtils.get(url, null);

            JsonElement jsonElement = JsonParser.parseString(jsonResponse);
            JsonObject jsonObject = jsonElement.getAsJsonObject();

            if (jsonObject.has("error")) {
                LOG.warn("Error from BE {}:{}: {}", host, port, jsonObject.get("error").getAsString());
                return profileFiles;
            }

            JsonArray profilesArray = jsonObject.getAsJsonArray("profiles");
            if (profilesArray == null) {
                return profileFiles;
            }

            SimpleDateFormat parseFormat = new SimpleDateFormat("yyyyMMdd-HHmmss");

            for (JsonElement element : profilesArray) {
                JsonObject profileObj = element.getAsJsonObject();
                String filename = profileObj.get("filename").getAsString();
                String type = profileObj.get("type").getAsString();
                String timestampStr = profileObj.get("timestamp").getAsString();
                long size = profileObj.get("size").getAsLong();

                try {
                    Date timestamp = parseFormat.parse(timestampStr);
                    profileFiles.add(new ProfileFileInfo(type, timestamp, size, filename));
                } catch (Exception e) {
                    LOG.warn("Failed to parse timestamp: {}", timestampStr, e);
                }
            }

            // Sort by timestamp descending (newest first)
            profileFiles.sort((f1, f2) -> f2.timestamp.compareTo(f1.timestamp));

        } catch (Exception e) {
            LOG.warn("Failed to fetch proc profiles from BE {}:{}", host, port, e);
        }

        return profileFiles;
    }

    private void appendProfileTableBody(StringBuilder buffer, List<ProfileFileInfo> profileFiles, String nodeParam) {
        buffer.append("<tbody>");
        SimpleDateFormat displayFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");

        for (ProfileFileInfo profile : profileFiles) {
            buffer.append("<tr>");
            buffer.append("<td>").append(profile.type).append("</td>");
            buffer.append("<td>").append(displayFormat.format(profile.timestamp)).append("</td>");
            buffer.append("<td>").append(DebugUtil.getPrettyStringBytes(profile.fileSize)).append("</td>");
            buffer.append("<td>");
            if ("FE".equals(nodeParam)) {
                buffer.append("<a href=\"/proc_profile/file?filename=")
                        .append(profile.fileName)
                        .append("\" target=\"_blank\">View</a>");
            } else {
                buffer.append("<a href=\"/proc_profile/file?node=")
                        .append(nodeParam)
                        .append("&filename=")
                        .append(profile.fileName)
                        .append("\" target=\"_blank\">View</a>");
            }
            buffer.append("</td>");
            buffer.append("</tr>");
        }
        buffer.append("</tbody>");
    }

    private static class ProfileFileInfo {
        final String type;
        final Date timestamp;
        final long fileSize;
        final String fileName;

        ProfileFileInfo(String type, Date timestamp, long fileSize, String fileName) {
            this.type = type;
            this.timestamp = timestamp;
            this.fileSize = fileSize;
            this.fileName = fileName;
        }
    }
}